File: cd/7.type checking/typeCheck.c
#include<stdio.h>
#define N 5
int main()
{
    int n, l, i, j, x, flag=0;
    char var[N], type[N], exp[N];
    char c;
    printf("\nEnter the no. of variable : ");
    scanf("%d",&n);
    for(i=0;i<n;i++)
    {
        printf("\nEnter the variable[%d] : ",i);
        scanf(" %c",&var[i]);
        printf("\nEnter the type of the variable ( float-f \t int-i ) : ");
        scanf(" %c",&type[i]);
        if(type[i]=='f') flag=1;
    }
    printf("\nEnter the expression end with $: ");
    c = getchar();
    i=0;
    while ((c=getchar())!='$')
    {
        exp[i]=c;
        i++;
        if(c=='/')
        {
            flag=1;
        }
    }
    l=i;
    for(i=0;i<l;i++)
    {
        for(j=0;j<n;j++)
        {
            if(exp[i]==var[j])
            {
                if(type[j]=='f')
                {
                    flag=1;
                }
            }
        }
    }
    for(i=0;i<n;i++)
    {
        if(exp[0]==var[i])
        {
            if(type[i]=='f')
            {
                printf("\nDatatype correctly defined!\n");
                break;
            }
            else
            {
                if(flag==1)
                {
                    printf("%c should be defined in float datatype!",var[i]);
                    break;
                }
                else
                {
                    printf("\nDatatype correctly defined!\n");
                    break;
                }
            }
        }
    }
    return 0;
}

File: cd/7.type checking/typeCheck.py
n = int(input("Enter the no. of variables : "))
dict = {}
flag = 0
for i in range(n):
    var = input(f"Enter the variable-{i+1} : ")
    type = input(f"Enter the type of the variable \t float-f \t int-i : ")
    dict[var] = type

expr = input("Enter the expression : ").split()
if '/' in expr:
    flag = 1

if dict[expr[0]]=='f':
    print("Data type correctly defined.!")
else:
    for i in range(2,len(expr)):
        if expr[i] not in ['+','-','*','/'] and dict[expr[i]]=='f':
            flag=1
            break
    if flag==1:
        print(f"{expr[0]} should be in float datatype.!")
    else:
        print("Data type correctly defined.!")
